#                                      Trabalho de Big Data - Nelson
<br>
<br>
Alunos:
<br>
    <ul>
        <li>
            Yuri Fontene Silva (1220105046)
        <li>
            Miguel de Moura Lobato (1220109077)
        </li>
        <li>
            João Victor Venancio Tavares (1220105097)
        </li>
    </ul>

## Dataset utilizado - Stores.xlsx
<br>
Link do dataset: https://www.kaggle.com/datasets/surajjha101/stores-area-and-sales-data
<br>
<br>
Contexto:
<p>
Este dataset foi retirado so site Kaggle.
    
No conjunto de dados, você obterá dados de diferentes lojas de uma empresa de supermercados de acordo com seus IDs de loja, que para facilitar foram convertidos em números inteiros positivos.
</p>
    <ul>
        <li>ID da loja: (Índice) ID da loja específica.</li>
        <li>Store_Area: Área física da loja em pátio.</li>
        <li>Items_Available: Número de itens diferentes disponíveis na loja correspondente.</li>
        <li>Daily_Customer_Count: Número de clientes que visitaram as lojas em média durante o mês.</li>
        <li>Store_Sales: Vendas em (US$) que as lojas realizaram.</li>
   </ul>

## Criando link com o dataset e mostrando em formato de Tabela

In [93]:
import pandas as pd
tabela = pd.read_excel('Stores.xlsx')
display(tabela)

ID  Store_Area  Items_Available  Daily_Customer_Count  Store_Sales
0      1        1659             1961                   530        66490
1      2        1461             1752                   210        39820
2      3        1340             1609                   720        54010
3      4        1451             1748                   620        53730
4      5        1770             2111                   450        46620
..   ...         ...              ...                   ...          ...
891  892        1582             1910                  1080        66390
892  893        1387             1663                   850        82080
893  894        1200             1436                  1060        76440
894  895        1299             1560                   770        96610
895  896        1174             1429                  1110        54340

[896 rows x 5 columns]

## Inserindo Coluna
<p>
    Criamos um função "imposto" que irá diminuir em 10% o valor que foi passao como parâmetro, para simular caso quiséssemos aplicar a taxação de um imposto em um relatório.
    Depois pegamos a mesma tabela que foi criada a cima e adcionamos a ela uma nova coluna chamada "Tax" para assim simular quanto pagamos de imposto.
    Logo abaixo criamos outra função que irá nos retornar o valor do 'Store_Sales' menos quanto iremos pagar de imposto. E assim criando outra coluna chamada "Post_Tax_Sales".
</p>

In [94]:
def imposto(valor):
    return round(valor * 0.10, 2)
def cal_imposto(valor):
    return round(valor - imposto(valor))
tabela['Tax'] = tabela['Store_Sales'].map(imposto)
tabela['Post_Tax_Sales'] = tabela['Store_Sales'].map(cal_imposto)
display(tabela)

ID  Store_Area  Items_Available  Daily_Customer_Count  Store_Sales  \
0      1        1659             1961                   530        66490   
1      2        1461             1752                   210        39820   
2      3        1340             1609                   720        54010   
3      4        1451             1748                   620        53730   
4      5        1770             2111                   450        46620   
..   ...         ...              ...                   ...          ...   
891  892        1582             1910                  1080        66390   
892  893        1387             1663                   850        82080   
893  894        1200             1436                  1060        76440   
894  895        1299             1560                   770        96610   
895  896        1174             1429                  1110        54340   

        Tax  Post_Tax_Sales  
0    6649.0           59841  
1    3982.0           35838  
2    5401.0           48609  
3    5373.0           48357  
4    4662.0           41958  
..      ...             ...  
891  6639.0           59751  
892  8208.0           73872  
893  7644.0           68796  
894  9661.0           86949  
895  5434.0           48906  

[896 rows x 7 columns]

## Ordenando Coluna

<p>
   Aqui nós chamamos mais uma vez o dataset para ser manipulado, criando uma nova variável que irá receber esse datatset porém com apenas as colunas "ID" e "Store_Sales". Após isso nós utilizamos da método 'sort_values' para ordenar em ordem decrescente apenas a coluna 'Store_Sales'.  
</p>

In [95]:
tabela_total = pd.read_excel('Stores.xlsx')
tabela_ordenada = tabela_total[['ID', 'Store_Sales']]
tabela_ordenada = tabela_ordenada.sort_values(by='Store_Sales', ascending=False)
display(tabela_ordenada)

ID  Store_Sales
649  650       116320
868  869       105150
432  433       102920
408  409       102310
758  759       101820
..   ...          ...
352  353        21300
593  594        20270
775  776        17670
852  853        16370
31    32        14920

[896 rows x 2 columns]

## Trabalhando com calculos estatíticos

<p>
     Criamos três diferentes variáveis para calcular a Média, Mediana e Moda. Depois criamos o DataFrame de cada variável para assim podermos transformá-las em uma tabela, onde elas seriam as colunas e as linhas seriam os informativos da tabela.
    
    OBS: Tivemos que utilizar o método '.iloc()' na variável 'modas' pois o método '.mode()' cria um array de modas, já que dentro de um datatset não existem apenas um tipo de valor que se repete.
</p>

In [96]:
colunas_selecionadas = ['Store_Area', 'Items_Available', 'Daily_Customer_Count', 'Store_Sales']

medias = tabela_total[colunas_selecionadas].mean()
medianas = tabela_total[colunas_selecionadas].median()
modas = tabela_total[colunas_selecionadas].mode().iloc[0]

df_medias = pd.DataFrame({'Médias': medias})
df_medianas = pd.DataFrame({'Medianas': medianas})
df_modas = pd.DataFrame({'Modas': modas})

tabela_medias_medianas_modas = pd.concat([df_medias, df_medianas, df_modas], axis=1)

display(tabela_medias_medianas_modas)

Médias  Medianas    Modas
Store_Area             1485.409598    1477.0   1439.0
Items_Available        1782.035714    1773.5   1858.0
Daily_Customer_Count    786.350446     780.0    670.0
Store_Sales           59351.305804   58605.0  54590.0

## Gráficos

<p>
   Foi importada a biblioteca 'ploty.express' para a criação de gráficos. Depois utilizamos do método '.bar' para criar um gráfico de barras com os seguintes parâmetros:
</p>
<br>
    tabela_ordenada: Este é o DataFrame ou conjunto de dados que você deseja usar para criar o gráfico. Parece que tabela_ordenada é o nome do DataFrame que contém seus dados.
    <br>
    <br>
    tabela_ordenada.index: Isso especifica a coluna que será usada para o eixo x do gráfico de barras. tabela_ordenada.index provavelmente se refere ao índice do DataFrame, que geralmente é uma série de rótulos de linha.
    <br>
    <br>
    y='Store_Sales': Este parâmetro especifica a coluna que será usada para o eixo y do gráfico de barras. No seu caso, parece que você está usando a coluna 'Store_Sales' para representar os valores das barras.


In [97]:
import plotly.express as px

grafico = px.bar(tabela_ordenada, tabela_ordenada.index, y='Store_Sales')
grafico.show()

<br>
<p>
   Este gráfico seguimos os mesmos passos do gráfico anterior. Porém dessa vez criamos mais uma coluna no eixo Y, dessa forma obtemos particamente o mesmo gráfico só que com algumas informações a mais sobre a coluna 'Items_Available'.
</p>
<br>

In [98]:
tabela_ordenada = tabela_total[['Store_Sales', 'Items_Available']]
grafico = px.bar(tabela_ordenada, tabela_ordenada.index, y=['Store_Sales', 'Items_Available'])
grafico.show()

## Problema identificado
<br>
<p>Ao analisar o nosso dataset identificamos que existem algumas lojas com uma discrepância de vendas, como exemplo as lojas de ID 868 e 32 respectivamente. Nelas nos podemos verificar que mesmo ambas tendo uma área e quantidade de produtos bem parecidas, ambas tiveram a mesma quantidade clientes no mês porém com valores de vendas bem distintos.</p><br><p> Podemos assim considerar a média de vendas da loja ID 32 igual a 15,07 enquanto a loja ID 869 teve 107,29. Um dos problemas que podemos observar é a quantidade de itens que a loja 868 tem a mais que a outra, isso pode acarretar em uma diversidade de itens mais abrangente trazendo assim mais clientes. Isso pode ser ocasionado por conta do espaço que a loja 32 tem (1250), o que leva a mesma a ter um estoque reduzido em relação a outra.</p>
<br>

In [99]:
tabela_total = pd.read_excel('Stores.xlsx')

colunas_selecionadas = ['ID', 'Store_Area', 'Items_Available', 'Daily_Customer_Count', 'Store_Sales']

primeira_linha = tabela_total[colunas_selecionadas].iloc[868]
ultima_linha = tabela_total[colunas_selecionadas].iloc[31]

tabela_resultado = pd.DataFrame([primeira_linha, ultima_linha])

display(tabela_resultado)

ID  Store_Area  Items_Available  Daily_Customer_Count  Store_Sales
868  869        1775             2104                   980       105150
31    32        1250             1508                   990        14920